diff --git a/aidl/als/AreaCapture.cpp b/aidl/als/AreaCapture.cpp
index 3c23c07..bc98dc8 100644
--- a/aidl/als/AreaCapture.cpp
+++ b/aidl/als/AreaCapture.cpp
@@ -23,6 +23,7 @@ using android::SurfaceComposerClient;
 using android::SyncScreenCaptureListener;
 using android::base::GetProperty;
 using android::gui::ScreenCaptureResults;
+using android::gui::SecureLayerMode;
 using android::gui::aidl_utils::toARect;
 using android::ui::PixelFormat;
 
@@ -59,7 +60,7 @@ ndk::ScopedAStatus AreaCapture::getAreaBrightness(AreaRgbCaptureResult* _aidl_re
     displayCaptureArgs.captureArgs.sourceCrop = toARect(m_screenshot_rect);
     displayCaptureArgs.width = m_screenshot_rect.getWidth();
     displayCaptureArgs.height = m_screenshot_rect.getHeight();
-    displayCaptureArgs.captureArgs.captureSecureLayers = true;
+    displayCaptureArgs.captureArgs.secureLayerMode = SecureLayerMode::Capture;
 
     sp<SyncScreenCaptureListener> captureListener = new SyncScreenCaptureListener();
     if (ScreenshotClient::captureDisplay(displayCaptureArgs, captureListener) !=

